### PR TITLE
TF-627: refactor mailbox dashbroad controller code and add advanced search code

### PR DESCRIPTION
**Pull request by ManhNTX**
Need merged [#30](https://github.com/linagora/jmap-dart-client/pull/30) before check this PR

### CODE REVIEW

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      Get.find<QuickSearchEmailInteractor>();
+
+  final accountId = Rxn<AccountId>();
+  final inMailbox = Rxn<MailboxId>();
+  final hasKeyword = Rxn<String>();
+  final notKeyword = Rxn<String>();
+  final hasAttachment = Rxn<bool>();
+  final patternSearch = Rxn<String>();
+  final from = RxList<String>();
+  final to = RxList<String>();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:31]: How about Strong type for this case `email`?

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/model/email_receive_time_type.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/model/quick_search_filter.dart';
+import 'package:tmail_ui_user/features/thread/domain/constants/thread_constants.dart';
+
+mixin SearchController {
+  final QuickSearchEmailInteractor _quickSearchEmailInteractor =
+      Get.find<QuickSearchEmailInteractor>();
+
+  final accountId = Rxn<AccountId>();
+  final inMailbox = Rxn<MailboxId>();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:25]: Manage all of these seem complicated. Other developer could make the mistake. Please find the way to keep in single place. Please think about this controller is a function `f` in math. it is better to minimize the number of variables the function `f` depend on. `f(x)` should be better than `f(x1, x2, x3, ....)`

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 3]
```diff
@@ -547,7 +547,8 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
                   controller.searchEmail(context, recent.value);
                 },
                 suggestionsCallback: (pattern) async {
-                  return controller.quickSearchEmailsAction(pattern);
+                  controller.patternSearch.value = pattern;
+                  return controller.searchEmailsAction();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:551]: I am not a fan of it. We should encourage functional programming (modify the `state` out of the function as little as possible

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:31]: IMO You should create an object containing all these properties for easy management. It is possible that we have many other properties to serve the search.

**Review by dab246 — CHANGES_REQUESTED** [review]

File: lib/features/mailbox_dashboard/presentation/model/advanced_search_filter.dart [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  to,
+  subject,
+  hasKeyword,
+  notKeyword,
+  mailBox,
+  date,
+  hasAttachment
+}
+
+extension AdvancedSearchFilterFieldExtension on AdvancedSearchFilterField {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/model/advanced_search_filter.dart:18]: please use `ehanced enum` of flutter v3

**Review by hoangdat — COMMENTED** [review]
commit `2d646b698f199066e1143e3d5b48ecc18878370d` create new empty file

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 5]
```diff
@@ -366,50 +369,55 @@ class MailboxDashBoardController extends ReloadableController {
...
+      consumeState(_deleteEmailPermanentlyInteractor.execute(
+          currentAccountId, email.id));
     }
   }
 
   void _deleteEmailPermanentlySuccess(DeleteEmailPermanentlySuccess success) {
     if (currentContext != null && currentOverlayContext != null) {
-      _appToast.showToastWithIcon(
-          currentOverlayContext!,
+      _appToast.showToastWithIcon(currentOverlayContext!,
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:354]: We prefer each parameter in separate line

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 6]
```diff
@@ -366,50 +369,55 @@ class MailboxDashBoardController extends ReloadableController {
...
+          _removeEmailDraftsInteractor.execute(currentAccountId, email.id));
     }
   }
 
   void deleteEmailPermanently(PresentationEmail email) {
     final currentAccountId = accountId.value;
     if (currentAccountId != null) {
-      consumeState(_deleteEmailPermanentlyInteractor.execute(currentAccountId, email.id));
+      consumeState(_deleteEmailPermanentlyInteractor.execute(
+          currentAccountId, email.id));
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:348]:
```
consumeState(
  _deleteEmailPermanentlyInteractor.execute(currentAccountId, email.id)
);
```

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 7]
```diff
@@ -445,11 +453,11 @@ class MailboxDashBoardController extends ReloadableController {
 
   UnsignedInt? get maxSizeAttachmentsPerEmail {
     if (sessionCurrent != null && accountId.value != null) {
-      final mailCapability = sessionCurrent
-          ?.accounts[accountId.value]
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:449]: no problem with it, please keep the current status

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 8]
```diff
@@ -462,47 +470,81 @@ class MailboxDashBoardController extends ReloadableController {
...
         .then((result) => result.fold(
             (failure) => <RecentSearch>[],
             (success) => success is GetAllRecentSearchLatestSuccess
                 ? success.listRecentSearch
                 : <RecentSearch>[]));
   }
 
-  Future<List<PresentationEmail>> quickSearchEmailsAction(String pattern) async {
+  Future<List<PresentationEmail>> quickSearchEmailsAction(
+      String pattern) async {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart]:
```
Future<List<PresentationEmail>> quickSearchEmailsAction(
    String pattern
) async {
```

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 9]
```diff
@@ -462,47 +470,81 @@ class MailboxDashBoardController extends ReloadableController {
...
+                      ? userProfile.value!.email
+                      : null,
+                  after: listFilterQuickSearch.contains(QuickSearchFilter.last7Days)
+                      ? UTCDate(
+                          DateTime.now().subtract(const Duration(days: 7)))
+                      : null,
+                  inMailbox: selectedMailbox.value!.id),
+              properties: ThreadConstants.propertiesQuickSearch)
+          .then((result) => result.fold((failure) => <PresentationEmail>[],
+              (success) => success is QuickSearchEmailSuccess ? success.emailList : <PresentationEmail>[]));
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart]: keep the current status

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 10]
```diff
@@ -462,47 +470,81 @@ class MailboxDashBoardController extends ReloadableController {
...
     }
     return <PresentationEmail>[];
   }
 
-  void selectQuickSearchFilter(QuickSearchFilter quickSearchFilter, {bool fromSuggestionBox = false}) {
+  Future<List<PresentationEmail>> advancedSearchEmailsAction(
+      String pattern) async {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart]: idem

**Review by hoangdat — COMMENTED** [review]
srr. It is still hard to me for reviewing :)

File: core/lib/presentation/views/bottom_popup/full_screen_action_sheet_builder.dart [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                appBar: AppBar(
+                  leading: SizedBox.shrink(),
+                  title: titleWidget,
+                  centerTitle: true,
+                  actions: [
+                    GestureDetector(
+                      child: cancelWidget,
+                      onTapDown: (_) {
+                        onCloseActionClick?.call();
+                        Navigator.pop(context);
```
Comment (dab246) [inline core/lib/presentation/views/bottom_popup/full_screen_action_sheet_builder.dart:60]: You should use `Get.back();`

File: env.file [hunk 12]
```diff
@@ -1,3 +1,3 @@
-SERVER_URL=http://localhost
+SERVER_URL=https://jmap.upn.integration-open-paas.org
```
Comment (dab246) [inline env.file:1]: You should only change this file locally

File: lib/features/login/domain/usecases/check_oidc_is_available_interactor.dart [hunk 13]
```diff
@@ -13,7 +13,7 @@ class CheckOIDCIsAvailableInteractor {
     try {
       final result = await _oidcRepository.checkOIDCIsAvailable(oidcRequest);
       log('CheckOIDCIsAvailableInteractor::execute(): result: $result');
-      return Right<Failure, Success>(CheckOIDCIsAvailableSuccess(result));
+      return Left<Failure, Success>(CheckOIDCIsAvailableFailure('dddd'));
```
Comment (dab246) [inline lib/features/login/domain/usecases/check_oidc_is_available_interactor.dart:16]: why change it?

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+  FocusNode searchFocus = FocusNode();
+
+  SearchEmailFilter get searchEmailFilterValue => searchEmailFilter.value;
+
+  SearchController(
+    this._quickSearchEmailInteractor,
+  );
+
+  selectOpenAdvanceSearch() {
+    isAdvancedSearchViewOpen.value = !isAdvancedSearchViewOpen.value;
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:65]: Should write `isAdvancedSearchViewOpen.toggle();`

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        ),
+        titleWidget: const Text(
+          'Advanced search',
+          style: TextStyle(
+            fontSize: 20,
+            color: AppColor.colorNameEmail,
+          ),
+        ),
+      ).show();
+      selectOpenAdvanceSearch();
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:99]: Why do you call it 2 times?

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+          padding: const EdgeInsets.only(right: 16),
+          child: SvgPicture.asset(
+            _imagePaths.icCloseAdvancedSearch,
+            color: AppColor.colorHintSearchBar,
+            width: 24,
+            height: 24,
+          ),
+        ),
+        titleWidget: const Text(
+          'Advanced search',
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:92]: You should add it to `AppLocalization`

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+  void registerSearchFocusListener() {
+    searchFocus.addListener(() {
+      final hasFocus = searchFocus.hasFocus;
+      final query = searchEmailFilter.value.text?.value;
+      log('MailboxDashBoardController::_registerSearchFocusListener(): hasFocus: $hasFocus | query: $query');
+      if (!hasFocus && (query == null || query.isEmpty)) {
+        updateFilterEmail(text: SearchQuery.initial());
+        searchInputController.clear();
+        cleanSearchFilter();
+        searchFocus.unfocus();      }
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:225]: sign `}`. should go down the line

File: lib/features/mailbox_dashboard/presentation/widgets/advanced_search_icon_widget.dart [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  final BuildContext _parentContext;
+
+  @override
+  Widget build(BuildContext context) {
+    return Obx(
+      () => buildIconWeb(
+          minSize: 0,
+          iconPadding: const EdgeInsets.only(right: 8),
+          icon: SvgPicture.asset(_imagePaths.icFilterAdvanced,
+              color: controller.isAdvancedSearchViewOpen.value
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/widgets/advanced_search_icon_widget.dart:28]: Should write `controller.isAdvancedSearchViewOpen.isTrue`

File: lib/features/mailbox_dashboard/presentation/widgets/advanced_search_input_form.dart [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+    Key? key,
+  }) : super(key: key);
+
+  final SearchController searchController;
+  final ImagePaths imagePaths;
+  final ResponsiveUtils responsiveUtils;
+
+  @override
+  Widget build(BuildContext context) {
+    // searchController.initSearchFilterField(context);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/widgets/advanced_search_input_form.dart]: remove it

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 20]
```diff
@@ -273,11 +252,13 @@ class MailboxDashBoardController extends ReloadableController {
 
   void setSelectedMailbox(PresentationMailbox? newPresentationMailbox) {
     selectedMailbox.value = newPresentationMailbox;
+    searchCtrl.updateFilterEmail(mailbox: selectedMailbox.value);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:255]: `SearchController` should observe that value from `dashboard` to follow single truth

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 21]
```diff
@@ -273,11 +252,13 @@ class MailboxDashBoardController extends ReloadableController {
...
   }
 
   void setNewFirstSelectedMailbox(PresentationMailbox? newPresentationMailbox) {
     selectedMailbox.firstRebuild = true;
     selectedMailbox.value = newPresentationMailbox;
+    searchCtrl.updateFilterEmail(mailbox: selectedMailbox.value);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:261]: idem

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 22]
```diff
@@ -246,6 +222,9 @@ class MailboxDashBoardController extends ReloadableController {
     if (arguments is Session) {
       sessionCurrent = arguments;
       accountId.value = sessionCurrent?.accounts.keys.first;
+      if(accountId.value != null){
+        searchCtrl.setAccountId(accountId.value!);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:226]: Single truth, should observe `accountId` from `dashboard`

**Pull request closed without merging**